pytorch_lightning/strategies/fully_sharded_native.py

```diff
@@ -0,0 +1,268 @@
+        based on your level of memory/speed efficiency. We suggest having a look at this tutorial for
+        more information.
+        `https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html`
+
+        Arguments:
+            cpu_offload (Optional [CPUOffload]):
+                CPU offloading config. Currently, only parameter and gradient CPU
+                offload is supported. It can be enabled via passing in
+                ``cpu_offload=CPUOffload(offload_params=True)``. Note that this
+                currently implicitly enables gradient offloading to CPU in order for
+                params and grads to be on same device to work with optimizer. This
+                API is subject to change. Default is ``None`` in which case there
+                will be no offloading.
+            backward_prefetch: (Optional[BackwardPrefetch]):
+                This is an experimental feature that is subject to change in the
+                the near future. It allows users to enable two different backward_prefetch
+                algorithms to help backward communication and computation overlapping.
+                Pros and cons of each algorithm is explained in the class ``BackwardPrefetch``.
+        """
+        if not _TORCH_GREATER_EQUAL_1_11:
+            raise MisconfigurationException(
+                "`DDPFullyShardedNativeStrategy` is supported from pytorch v1.11.0 onwards."
+            )
+
+        super().__init__(
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+        self._process_group = None
+        self.num_processes = len(self.parallel_devices) if self.parallel_devices is not None else 0
+        self._process_group_backend: Optional[str] = process_group_backend
+        self.cpu_offload: Optional[CPUOffload] = cpu_offload
+        self.backward_prefetch: Optional[BackwardPrefetch] = backward_prefetch
+
+    @property
+    def root_device(self) -> torch.device:
+        return self.parallel_devices[self.local_rank]
+
+    @property
+    def process_group(self) -> Optional[ProcessGroup]:
+        if self._process_group is None:
+            # The plugin should have already initilized ddp in setup_environment()
+            self._process_group = _get_default_group()
+        return self._process_group
+
+    @property
+    def setup_optimizers_in_pre_dispatch(self) -> bool:
+        # Setup optimizers after the Fully Sharded Model has been made
+        return True
+
+    @property
+    def process_group_backend(self) -> Optional[str]:
+        return self._process_group_backend
+
+    def setup_environment(self) -> None:
+        self.setup_distributed()
+        super().setup_environment()
+
+    def setup_distributed(self) -> None:
```

Comment:
move this logic directly into setup_environment. we don't need both `setup_environment` and `setup_distributed` as top-level, public methods. the strategy API only offers `setup_environment`